File: Parallelism.vb
Imports System.Threading

Module Parallelism
    Sub Main()
        Dim threads = 5
        Dim count = 0
        Dim lobj = New Object()
        Dim listener As Action = Function()
                                     SyncLock lobj
                                         count += 1
                                         Thread.Sleep(1000)
                                         Console.WriteLine(count)
                                     End SyncLock
                                 End Function

        For i = 1 To threads
            Dim t = New Thread(AddressOf doit)
            t.Start(listener)
        Next
    End Sub

    Sub doit(cb As Object)
        ' check cpu parallelism
        'While True
        'End While

        Thread.Sleep(1000)
        Dim _cb = CType(cb, Action)
        Console.WriteLine("並行")
        _cb()
    End Sub
End Module

File: Parallelism.cs
using System;
using System.Threading;

namespace cscs
{
    class Parallelism
    {
        static void Main(string[] args)
        {            
            var lobj = new Object();
            var threads = 5;
            var count = 0;
            Action listener = () => {
                lock (lobj)
                {
                    ++count;
                    Thread.Sleep(1000);
                    Console.WriteLine(count);
                }
            };

            for (var i = 0; i < threads; ++i)
            {
                var t = new Thread(doit);
                t.Start(listener);
            }
        }

 
        static void doit(Object cb)
        {
            /* check cpu parallelism
            while (true){ }
            //*/

            Thread.Sleep(1000);
            var _cb = cb as Action;
            Console.WriteLine("並行");
            _cb();
        }
    }
}
